File: Master_the_Key_ROS_Concepts/Lec1_Create_a_Cakin_Workspace/catkin_ws/src/battery_status/src/battery_node.cpp
#include <ros/ros.h>
#include <battery_msgs/Battery_Srv.h>
//#include <cstdlib>
//int state = 1;

#define STATE_FULL 1
#define STATE_EMPTY 2


void set_led(ros::NodeHandle *nh, int battery_state)
{
	ros::service::waitForService("/set_led");
	
	ros::ServiceClient client = nh->serviceClient<battery_msgs::Battery_Srv>("/set_led");

	battery_msgs::Battery_Srv srv;
	srv.request.led_number = 1;
	srv.request.state = 0;

	if(battery_state==STATE_EMPTY)
	{
		srv.request.state = 1;
	}

	if(client.call(srv)){
		ROS_INFO("Set led success flag: %d", srv.response.success);
	}
	else{
		ROS_WARN("Service call failed");
	}
}

int main(int argc, char **argv)
{
	ros::init(argc, argv, "battery_node");
	ros::NodeHandle nh;

	int battery_state = STATE_FULL;
	
	while (ros::ok()){
		ros::Duration(7.0).sleep();
		battery_state = STATE_EMPTY;
		ROS_INFO("Battery is empty!");
		set_led(&nh, battery_state);
		ros::Duration(3.0).sleep();
		battery_state = STATE_FULL;
		ROS_INFO("Battery is full!");
		set_led(&nh, battery_state);
	}
}

File: Master_the_Key_ROS_Concepts/Lec1_Create_a_Cakin_Workspace/catkin_ws/src/battery_status/scripts/battery_node.py
#!/usr/bin/env python3

import rospy
from battery_msgs.srv import Battery_Srv

def set_led(battery_state):
    rospy.wait_for_service("/set_led")
    try:
        service = rospy.ServiceProxy("/set_led", Battery_Srv)
        state = 0
        if battery_state == 'empty':
            state = 1
        resp = service(1, state)
        rospy.loginfo("Set led success flag: " + str(resp))
    
    except rospy.ServiceException as e:
        rospy.logerr(e)

if __name__ == '__main__':
    rospy.init_node('battery_node')

    battery_state = "full"

    while not rospy.is_shutdown():
        rospy.sleep(7)
        battery_state = "empty"
        rospy.loginfo("The battery is empty!")
        set_led(battery_state)
        rospy.sleep(3)
        battery_state = "full"    
        rospy.loginfo("The battery is full!")
        set_led(battery_state)
